### ISSUE TITLE
Trouble with Triplet Loss Implementation in MLX

### ISSUE THREAD

**Issue by sarim-zafar** [ml-explore/mlx-examples#659]
**Issue Description:**

I attempted to port the triplet loss implementation from PyTorch to MLX (machine learning extension for Apache MXNet) using the provided reference code as a guide. However, when I attempted to train a simple MLP on MNIST data using the implemented triplet loss function, I encountered unexpected behavior. Specifically, the loss quickly falls very close to zero during training, and the resulting embeddings seem to be worse than random initialization.

**Steps to Reproduce:**

1. Ported the triplet loss implementation from PyTorch to MLX.
2. Implemented a simple MLP model for training on MNIST data.
3. Used the MLX triplet loss function during training.
4. Observed that loss quickly approaches zero, and resulting embeddings are suboptimal.

**Code Snippet Used:**

```
import mlx.core as mx
import mlx.nn as nn


def euclidean_distance_matrix(x,eps = 1e-8):
  """Efficient computation of Euclidean distance matrix
  Args:
    x: Input tensor of shape (batch_size, embedding_dim)
    
  Returns:
    Distance matrix of shape (batch_size, batch_size)
  """
  # step 1 - compute the dot product

  # shape: (batch_size, batch_size)
  dot_product = mx.matmul(x, x.transpose())

  # step 2 - extract the squared Euclidean norm from the diagonal

  # shape: (batch_size,)
  squared_norm = mx.diag(dot_product)

  # step 3 - compute squared Euclidean distances

  # shape: (batch_size, batch_size)
  distance_matrix = (mx.expand_dims(squared_norm, axis=0) - 2 * dot_product +
                      mx.expand_dims(squared_norm, axis=1))

  # get rid of negative distances due to numerical instabilities
  distance_matrix = nn.relu(distance_matrix)

  # step 4 - compute the non-squared distances
  
  # handle numerical stability
  # derivative of the square root operation applied to 0 is infinite
  # we need to handle by setting any 0 to eps
  mask = (distance_matrix == 0.0).astype(mx.float32)

  # use this mask to set indices with a value of 0 to eps
  distance_matrix += mask * eps

  # now it is safe to get the square root
  distance_matrix = mx.sqrt(distance_matrix)

  # undo the trick for numerical stability
  distance_matrix *= (1.0 - mask)

  return distance_matrix



def get_triplet_mask(labels):
  """compute a mask for valid triplets
  Args:
    labels: Batch of integer labels. shape: (batch_size,)
  Returns:
    Mask tensor to indicate which triplets are actually valid. Shape: (batch_size, batch_size, batch_size)
    A triplet is valid if:
    `labels[i] == labels[j] and labels[i] != labels[k]`
    and `i`, `j`, `k` are different.
  """
  # step 1 - get a mask for distinct indices

  # shape: (batch_size, batch_size)
  indices_equal = mx.eye(labels.shape[0], dtype=mx.bool_)

  indices_not_equal = mx.logical_not(indices_equal)
  # shape: (batch_size, batch_size, 1)
  i_not_equal_j = mx.expand_dims(indices_not_equal, axis=2)
  # shape: (batch_size, 1, batch_size)
  i_not_equal_k = mx.expand_dims(indices_not_equal, axis=1)
  # shape: (1, batch_size, batch_size)
  j_not_equal_k = mx.expand_dims(indices_not_equal, axis=0)
  # Shape: (batch_size, batch_size, batch_size)
  distinct_indices = mx.logical_and(mx.logical_and(i_not_equal_j, i_not_equal_k), j_not_equal_k)

  # step 2 - get a mask for valid anchor-positive-negative triplets

  # shape: (batch_size, batch_size)
  labels_equal = mx.expand_dims(labels, axis=0) == mx.expand_dims(labels, axis=1)
  # shape: (batch_size, batch_size, 1)
  i_equal_j = mx.expand_dims(labels_equal, axis=2)
  # shape: (batch_size, 1, batch_size)
  i_equal_k = mx.expand_dims(labels_equal, axis=1)
  # shape: (batch_size, batch_size, batch_size)
  valid_indices = mx.logical_and(i_equal_j, mx.logical_not(i_equal_k))

  # step 3 - combine two masks
  mask = mx.logical_and(distinct_indices, valid_indices)

  return mask


def BatchAllTripletLoss(embeddings, labels, margin=1.0, eps=1e-8):
    """Uses all valid triplets to compute Triplet loss
    Args:
        embeddings: Batch of embeddings, e.g., output of the encoder. shape: (batch_size, embedding_dim)
        labels: Batch of integer labels associated with embeddings. shape: (batch_size,)
        margin: Margin value in the Triplet Loss equation computes loss value.
        eps: Small epsilon value for numerical stability.

    Returns:
        Scalar loss value.
    """
    # Step 1 - get distance matrix
    distance_matrix = euclidean_distance_matrix(embeddings, eps=eps)

    # Step 2 - compute loss values for all triplets by applying broadcasting to distance matrix
    anchor_positive_dists = mx.expand_dims(distance_matrix, axis=2)
    anchor_negative_dists = mx.expand_dims(distance_matrix, axis=1)
    triplet_loss = anchor_positive_dists - anchor_negative_dists + margin

    # Step 3 - filter out invalid or easy triplets by setting their loss values to 0
    mask = get_triplet_mask(labels)
    triplet_loss *= mask
    triplet_loss = nn.relu(triplet_loss)

    # Step 4 - compute scalar loss value by averaging positive losses
    num_positive_losses = mask.sum()  # Count the number of positive losses
    triplet_loss = triplet_loss.sum() / (num_positive_losses + eps)

    return triplet_loss


def TripletSemiHardLoss(embeddings, labels, margin=1.0,eps=1e-8):
    """Computes the triplet loss_functions with semi-hard negative mining.
       The loss_functions encourages the positive distances (between a pair of embeddings
       with the same labels) to be smaller than the minimum negative distance
       among which are at least greater than the positive distance plus the
       margin constant (called semi-hard negative) in the mini-batch.
       If no such negative exists, uses the largest negative distance instead.
       See: https://arxiv.org/abs/1503.03832.
       We expect labels `labels` to be provided as 1-D integer `array` with shape
       [batch_size] of multi-class integer labels. And embeddings `embeddings` must be
       2-D float `Tensor` of l2 normalized embedding vectors.
       Args:
         margin: Float, margin term in the loss_functions definition. Default value is 1.0.
         name: Optional name for the op.
       """

    # Reshape label tensor to [batch_size, 1].
    lshape = labels.shape
    labels = mx.reshape(labels, [lshape[0], 1])

    pdist_matrix = euclidean_distance_matrix(embeddings,eps = eps)

    # Build pairwise binary adjacency matrix.
    adjacency = mx.array_equal(labels, labels.transpose(0, 1))
    # Invert so we can select negatives only.
    adjacency_not = mx.logical_not(adjacency)

    batch_size = labels.shape[0]

    # Compute the mask.
    pdist_matrix_tile = mx.tile(pdist_matrix,[batch_size, 1])
    adjacency_not_tile = mx.tile(adjacency_not,[batch_size, 1])

    transpose_reshape = pdist_matrix.transpose(0, 1).reshape(-1, 1)
    greater = pdist_matrix_tile > transpose_reshape

    mask = adjacency_not_tile & greater

    # final mask
    mask_step = mask.astype(mx.float32)
    mask_step = mask_step.sum(axis=1)
    mask_step = mask_step > 0.0
    mask_final = mask_step.reshape(batch_size, batch_size)
    mask_final = mask_final.transpose(0, 1)

    adjacency_not = adjacency_not.astype(mx.float32)
    mask = mask.astype(mx.float32)

    # negatives_outside: smallest D_an where D_an > D_ap.
    axis_maximums = mx.max(pdist_matrix_tile, dim=1, keepdim=True)
    masked_minimums = mx.min(mx.mul(pdist_matrix_tile - axis_maximums[0], mask), dim=1, keepdim=True)[0] + axis_maximums[0]
    negatives_outside = masked_minimums.reshape([batch_size, batch_size])
    negatives_outside = negatives_outside.transpose(0, 1)

    # negatives_inside: largest D_an.
    axis_minimums = mx.min(pdist_matrix, dim=1, keepdim=True)
    masked_maximums = mx.max(mx.mul(pdist_matrix - axis_minimums[0], adjacency_not), dim=1, keepdim=True)[0] + axis_minimums[0]
    negatives_inside = masked_maximums.repeat(1, batch_size)

    semi_hard_negatives = mx.where(mask_final, negatives_outside, negatives_inside)

    loss_mat = margin + pdist_matrix - semi_hard_negatives

    mask_positives = adjacency.astype(mx.float32) - mx.diag(mx.ones(batch_size))
    num_positives = mask_positives.sum()

    triplet_loss = (mx.max(mx.mul(loss_mat, mask_positives), mx.tensor([0.]))).sum() / num_positives
    triplet_loss = triplet_loss.astype(mx.float32)
    return triplet_loss
```

**Expected Behavior:**

The triplet loss function should facilitate effective training of the MLP model, resulting in meaningful embeddings that capture the data structure.

**Actual Behavior:**

The loss quickly approaches zero, and the resulting embeddings are of poor quality, potentially worse than random initialization.

**Additional Context:**

I referred to the following code as a guide for implementing the triplet loss function:
[Pytorch Triplet Loss with semi hard mining](https://github.com/alfonmedela/triplet-loss-pytorch/blob/master/loss_functions/triplet_loss.py)

**Request for Assistance:**

I'm seeking guidance on identifying potential issues with the triplet loss implementation or suggestions for improving the performance of the embeddings. Any insights or assistance would be greatly appreciated.

**Environment Details:**

- MLX version: 0.7
- Python version: 3.9
- Operating System: Mac OS 14.0

**Note:** I'm unsure if this is the appropriate forum for addressing this issue, but any help or pointers in the right direction would be immensely valuable.

--- 

Looking forward to any assistance or advice the community can provide. Thank you!

**Comment by awni** [comment]
What do you expect the loss to go to?

As a debugging step I would compare the same loss in PyTorch and MLX with the same data (MNIST) and same model (MLP).

Also I would use the latest MLX (0.9.1).

**Comment by awni** [comment]
Closing as inactive. Feel free to comment if you need further assistance.
